src/coreclr/jit/ee_il_dll.hpp

```diff
@@ -78,9 +78,18 @@ bool Compiler::eeIsFieldStatic(CORINFO_FIELD_HANDLE fldHnd)
 }
 
 FORCEINLINE
-var_types Compiler::eeGetFieldType(CORINFO_FIELD_HANDLE fldHnd, CORINFO_CLASS_HANDLE* pStructHnd)
+var_types Compiler::eeGetFieldType(CORINFO_FIELD_HANDLE  fldHnd,
+                                   CORINFO_CLASS_HANDLE* pStructHnd,
+                                   CORINFO_CLASS_HANDLE  memberParent)
 {
-    return JITtype2varType(info.compCompHnd->getFieldType(fldHnd, pStructHnd));
+    // memberParent is an opportunistic hint to get a more exact field type
+    if ((memberParent != NO_CLASS_HANDLE) &&
+        !info.compCompHnd->isMoreSpecificType(info.compCompHnd->getFieldClass(fldHnd), memberParent))
```

Comment:
This looks suspect. What's the problem with passing the hint to `getFieldType` unconditionally?

Reply:
> This looks suspect. What's the problem with passing the hint to `getFieldType` unconditionally?

`getFieldType` uses that argument for validation purposes and requires it to be actual owner or its subclass. I've added a validation loop in my latest commits - I presume we can also modify JIT-EE interface and duplicate the same loop in both runtimes so the jit side change looks better?

Reply:
What are the cases when we reject the hint? Can the hint be completely bogus type; or is the worse-case scenario that the hint is `Canon` instantiation instead of actual generic instantiation?

Reply:
> What are the cases when we reject the hint? Can the hint be completely bogus type; or is the worse-case scenario that the hint is `Canon` instantiation instead of actual generic instantiation?

For the hint we call our main work-horse `gtGetClassHandle(GenTree*)` for an object representing an instance of the class with the field, so it can report something generic like `System.Object` or even an interface so without any validation `getFieldType` will fail.

So the loop I pushed will make sure the hinted type actually contains the field and `isMoreSpecific` is used just to ignore cases where `gtGetClassHandle(GenTree*)` fails to determine the exact class of the field owner.

Reply:
Field handles are shared between Canon and specific instantiations in the CoreCLR runtime type system and only for instance fields. I do not think that the validation will work as expected in AOT compilers. 

The validation is very expensive compared to what it does. The same check can be 100x cheaper if done on the VM side directly.

Also, is the current implementation leaving some cases on the table? Consider this:
```
class A<T> { T _aField; }
class B : A<string> { }
```
Is it possible to end up with `B` as the hint type for `_aField`?

Reply:
@jkotas I ended up moving logic to `getTypeFIeld` VM API. It looks like the 3rd argument was only used in 2 places (I think both were written by me) so it looks like it's fine to relax the argument to a potentially an invalid hint

Reply:
Looks good to me.